stumpy/stumpi.py

```diff
@@ -229,28 +247,34 @@ def _update_egress(self, t):
 
         core.apply_exclusion_zone(D, D.shape[0] - 1, self._excl_zone, np.inf)
 
-        update_idx = np.argwhere(D < self._P).flatten()
-        self._I[update_idx] = D.shape[0] + self._n_appended - 1  # D.shape[0] is base-1
-        self._P[update_idx] = D[update_idx]
+        update_idx = np.argwhere(D < self._P[:, -1]).flatten()
+        for i in update_idx:
+            pos = np.searchsorted(self._P[i], D[i], side="right")
+            core._shift_insert_at_index(self._P[i], pos, D[i])
+            core._shift_insert_at_index(
+                self._I[i], pos, D.shape[0] + self._n_appended - 1
+            )
+            # D.shape[0] is base-1
 
-        I_last = np.argmin(D)
+        # O(Nlog(K)) time complexity
```

Comment:
So, just to let others know about the time complexity and see if you/others have better idea. Personally speaking, I think that's it and we cannot do better. (I think I should remove it)

Reply:
I should have written better: 
```
# `self._P[-1, -1]` is the largest value of   `self._P[-1]`
```

Just something to help reader understand why we do `d < self._P[-1, -1]`

Reply:
It's pretty good. When `k` is small, this is essentially linear time `O(n)`. It might be worth posting a general question to Stackoverflow to see if people have ideas for improving it.

Reply:
Sure. Will do so. Also, I will remove the comment.